File: NinePSharp.Messages.FSharp/messages/Rread.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rread =
    { Size: uint32
      Tag: uint16
      Count: uint32
      Data: byte[] }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rread
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rread span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Count); offset <- offset + 4
            if not (isNull this.Data) then
                this.Data.CopyTo(span.Slice(offset))

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let count = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let dataArr = data.Slice(offset, int count).ToArray()
        { Size = size; Tag = tag; Count = count; Data = dataArr }

File: NinePSharp/messages/Rread.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Rread tag[2] count[4] data[count]
public readonly struct Rread : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rread;
    public ushort Tag { get; }

    public uint Count { get; }
    public ReadOnlyMemory<byte> Data { get; }

    public Rread(ushort tag, ReadOnlyMemory<byte> data)
    {
        Tag = tag;
        Data = data;
        Count = (uint)Data.Length;
        Size = (uint)(NinePConstants.HeaderSize + 4 + Data.Length);
    }

    public Rread(ReadOnlyMemory<byte> data)
    {
        var span = data.Span;
        Size = BinaryPrimitives.ReadUInt32LittleEndian(span[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(span.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;

        Count = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(offset, 4));
        offset += 4;

        // Zero-copy: Reference the existing memory slice
        Data = data.Slice(offset, (int)Count);
    }
    
    public Rread(uint size, ushort tag, uint count, ReadOnlyMemory<byte> data)
    {
        Size = size;
        Tag = tag;
        Count = count;
        Data = data;
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Count);
        offset += 4;

        Data.Span.CopyTo(data.Slice(offset, (int)Count));
    }
}
